pysb/simulator/base.py

```diff
@@ -0,0 +1,329 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+import itertools
+import sympy
+from pysb.core import MonomerPattern, ComplexPattern, as_complex_pattern
+
+
+class SimulatorException(Exception):
+    pass
+
+
+class Simulator(object):
+    """An abstract base class for numerical simulation of models.
+
+    Parameters
+    ----------
+    model : pysb.Model
+        Model to simulate.
+    tspan : vector-like, optional
+        Time values over which to simulate. The first and last values define
+        the time range. Returned trajectories are sampled at every value unless
+        the simulation is interrupted for some reason, e.g., due to
+        satisfaction
+        of a logical stopping criterion (see 'tout' below).
+    verbose : bool, optional (default: False)
+        Verbose output.
+
+    Attributes
+    ----------
+    verbose: bool
+        Verbosity flag passed to the constructor.
+    model : pysb.Model
+        Model passed to the constructor.
+    tspan : vector-like
+        Time values passed to the constructor.
+    tout: numpy.ndarray
+        Time points returned by the simulator (may be different from ``tspan``
+        if simulation is interrupted for some reason).
+
+    Notes
+    -----
+    If ``tspan`` is not defined, it may be defined in the call to the
+    ``run`` method.
+
+    The dimensionality of ``tout`` depends on whether a single simulation
+    or multiple simulations are run.
+
+    The dimensionalities of ``y``, ``yobs``, ``yobs_view``, ``yexpr``, and
+    ``yexpr_view`` depend on the number of simulations being run as well
+    as on the type of simulation, i.e., spatial vs. non-spatial.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def __init__(self, model, verbose=False, **kwargs):
+        self.model = model
+        self.verbose = verbose
+        self.tout = None
+        self._y = None
```

Comment:
Maybe just call self._clear_outputs instead of duplicating all those `= None` lines here? _clear_outputs might better be renamed to _init_outputs then.